substrate/frame/staking-async/rc-client/src/lib.rs

```diff
@@ -287,6 +331,137 @@ impl<AccountId> SessionReport<AccountId> {
+/// the correct call indices, which we only know at the runtime level.
+pub struct XCMSender<Sender, Destination, Message, ToXcm>(
+	core::marker::PhantomData<(Sender, Destination, Message, ToXcm)>,
+);
+
+impl<Sender, Destination, Message, ToXcm> XCMSender<Sender, Destination, Message, ToXcm>
+where
+	Sender: SendXcm,
+	Destination: Get<Location>,
+	Message: SplittableMessage + Display + Clone + Encode,
+	ToXcm: Convert<Message, Xcm<()>>,
+{
+	/// Safe send method to send a `message`, while validating it and using [`SplittableMessage`] to
+	/// split it into smaller pieces if XCM validation fails with `ExceedsMaxMessageSize`. It will
+	/// fail on other errors.
+	///
+	/// It will only emit some logs, and has no return value. This is used in the runtime, so it
+	/// cannot deposit any events at this level.
+	pub fn split_then_send(message: Message, maybe_max_steps: Option<u32>) {
+		let message_type_name = core::any::type_name::<Message>();
+		let dest = Destination::get();
+		let xcms = match Self::prepare(message, maybe_max_steps) {
+			Ok(x) => x,
+			Err(e) => {
+				log::error!(target: "runtime::rc-client", "📨 Failed to split message {}: {:?}", message_type_name, e);
+				return;
+			},
+		};
+
+		for (idx, xcm) in xcms.into_iter().enumerate() {
+			log::debug!(target: "runtime::rc-client", "📨 sending {} message index {}, size: {:?}", message_type_name, idx, xcm.encoded_size());
+			let result = send_xcm::<Sender>(dest.clone(), xcm);
+			match result {
+				Ok(_) => {
+					log::debug!(target: "runtime::rc-client", "📨 Successfully sent {} message part {} to relay chain", message_type_name,  idx)
+				},
+				Err(e) => {
+					log::error!(target: "runtime::rc-client", "📨 Failed to send {} message to relay chain: {:?}", message_type_name, e)
+				},
+			}
+		}
+	}
+
+	fn prepare(message: Message, maybe_max_steps: Option<u32>) -> Result<Vec<Xcm<()>>, SendError> {
+		// initial chunk size is the entire thing, so it will be a vector of 1 item.
+		let mut chunk_size = message.len();
+		let mut steps = 0;
+
+		loop {
+			let current_messages = message.clone().split_by(chunk_size);
+
+			// the first message is the heaviest, the last one might be smaller.
+			let first_message = if let Some(r) = current_messages.first() {
+				r
+			} else {
+				log::debug!(target: "runtime::staking-async::xcm", "📨 unexpected: no messages to send");
+				return Ok(vec![]);
+			};
+
+			log::debug!(
+				target: "runtime::staking-async::xcm",
+				"📨 step: {:?}, chunk_size: {:?}, message_size: {:?}",
+				steps,
+				chunk_size,
+				first_message.encoded_size(),
+			);
+
+			let first_xcm = ToXcm::convert(first_message.clone());
+			match <Sender as SendXcm>::validate(&mut Some(Destination::get()), &mut Some(first_xcm))
+			{
+				Ok((_ticket, price)) => {
+					log::debug!(target: "runtime::staking-async::xcm", "📨 validated, price: {:?}", price);
+					return Ok(current_messages.into_iter().map(ToXcm::convert).collect::<Vec<_>>());
```

Comment:
I'm not sure if I understand this correctly. At first, you're trying to validate whether it's okay to send just one message using `first_xcm = ToXcm::convert(current_messages.first())`. If the validation succeeds, you then return all the current_messages, not just the one that was validated?
I think this isn't fully compliant with the SendXcm trait's description. According to it, the ticket returned by SendXcm::validate should be passed to SendXcm::deliver.

Or is this kind of "best effort" strategy? Validate at least first one and then try to send more and just log erros when they fail?